ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/BackingStore/LMDB/TransactionHandle.hs

```diff
@@ -0,0 +1,135 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE Rank2Types          #-}
+{-# LANGUAGE RoleAnnotations     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+{- | Transactions handles.
+
+  There are two downsides to regular @'Transaction'@s:
+  * The current @'Transaction'@ runners (like @'readWriteTransaction'@) will
+    create and destroy an internal transaction handle (@'MDB_txn'@), which means
+    that we can not use the fact that the transaction handle provides a
+    consistent view of database if we want to perform several transactions to
+    the same consistent view.
+  * A @'Transaction'@ will only produce results when it is run with
+    @'readWriteTransaction'@ or one of the other runners. This prevents us from
+    getting intermediate results or performing intermediate writes (side
+    effects).
+
+  A @'TransactionHandle'@ will keep open an internal LMDB transaction handle
+  (@'MDB_txn'@) in an @'MVar'@, such that we can submit the usual
+  @'Transaction'@s to it. Each submission will provide its results and peform
+  its side effects right away.
+
+  Note on terminology: whenever we speak of a transaction handle, we refer to a
+  @'TransactionHandle'@, unless we say it is an /internal (LMDB)/ transaction
+  handle, in which case we refer to an @'MDB_txn'@.
+
+  TODO: upstream this module to @lmdb-simple@.
+-}
+module Ouroboros.Consensus.Storage.LedgerDB.BackingStore.LMDB.TransactionHandle (
+    TransactionHandle
+  , commit
+  , new
+  , newReadOnly
+  , newReadWrite
+  , submit
+  , submitReadOnly
+  , submitReadWrite
+  ) where
+
+import           Control.Concurrent.MVar
+import           Control.Monad
+
+import           Database.LMDB.Raw
+import           Database.LMDB.Simple.Internal
+
+type role TransactionHandle nominal
+
+-- | A @'TransactionHandle'@ will keep open an internal LMDB transaction handle.
+newtype TransactionHandle mode = TransactionHandle {
```

Comment:
Check:
- Open read
- Open write and do some writes
- Open another read
Does it see the writes?

Reply:
We currently can't test this because read-write transaction handles are not implemented. It would be a good check to do though, when/if we can also use read-write transaction handles!